subsys/bluetooth/audio/csip_set_coordinator.c

```diff
@@ -211,33 +213,65 @@ static int member_rank_compare_asc(const void *m1, const void *m2)
 	return svc_inst_1->set_info->rank - svc_inst_2->set_info->rank;
 }
 
-static int member_rank_compare_desc(const void *m1, const void *m2)
+static int member_rank_compare_desc(const struct bt_csip_set_coordinator_set_member *member_1,
+				    const struct bt_csip_set_coordinator_set_member *member_2,
+				    const struct bt_csip_set_coordinator_set_info *info)
 {
 	/* If we call the "compare ascending" function with the members
 	 * reversed, it will work as a descending comparison
 	 */
-	return member_rank_compare_asc(m2, m1);
+	return member_rank_compare_asc(member_2, member_1, info);
 }
 
 static void active_members_store_ordered(const struct bt_csip_set_coordinator_set_member *members[],
 					 size_t count,
 					 const struct bt_csip_set_coordinator_set_info *info,
 					 bool ascending)
 {
+	/* FIXME: The operation below shall not be allowed, as it looses pointer constness */
 	(void)memcpy(active.members, members, count * sizeof(members[0U]));
 	active.members_count = count;
 	active.info = info;
 
+	bt_csip_set_coordinator_set_members_sort_by_rank(
+		(const struct bt_csip_set_coordinator_set_member **)active.members,
+		active.members_count, active.info, ascending);
+}
+
+void bt_csip_set_coordinator_set_members_sort_by_rank(
+	const struct bt_csip_set_coordinator_set_member *members[], size_t count,
+	const struct bt_csip_set_coordinator_set_info *info, bool ascending)
+{
+	member_rank_compare_func_t func = ascending ? member_rank_compare_asc :
+						      member_rank_compare_desc;
+
 	if (count > 1U && CONFIG_BT_MAX_CONN > 1) {
-		qsort(active.members, count, sizeof(members[0U]),
-		ascending ? member_rank_compare_asc : member_rank_compare_desc);
+		for (size_t i = 0; i < count - 1; i++) {
+			bool swapped = false;
+
+			for (size_t j = 0; j < count - i - 1; j++) {
+				if (func(members[j], members[j + 1], info) > 0) {
+					const struct bt_csip_set_coordinator_set_member *tmp;
+
+					tmp = members[j];
+					members[j] = members[j + 1];
+					members[j + 1] = tmp;
+
+					swapped = true;
+				}
+			}
+
+			if (!swapped) {
+				break;
+			}
+		}
```

Comment:
What is the advantage of using bubblesort over qsort, since you replaced it?

Reply:
It's because `qsort` does not allow to pass `user_data`, like `struct bt_csip_set_coordinator_set_info *` in this case.
As the existing `active_members_store_ordered` stores the `info` in global `active` context variable, I could not call this function. Thus I added custom sorting function that has `const struct bt_csip_set_coordinator_set_info *info` as one of the parameters and can be used outside as well.

Reply:
Can't we use `qsort_r` for this instead?
From the `qsort` manpage:

> The qsort_r() function is identical to qsort() except that the comparison function compar takes  a  third argument. A pointer is passed to the comparison function via arg.  In this way, **the comparison function does not need to use global variables to pass through arbitrary arguments**, and is therefore reentrant and safe to use in threads.